src/modules/hip/kernel/tensor_mean.hpp

```diff
@@ -0,0 +1,695 @@
+#include <hip/hip_runtime.h>
+#include "rpp_hip_common.hpp"
+#include "reduction.hpp"
+
+// -------------------- Set 0 - Reduction Stage 2 --------------------
+__global__ void tensor_mean_grid_result_hip(Rpp32u *srcPtr,
+                                            uint xBufferLength,
+                                            float *dstPtr,
+                                            RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = hipThreadIdx_x * 8;
+    int id_z = hipBlockIdx_z;
+
+    __shared__ uint partialSum_smem[1024];                          // 8192 uints of src reduced to 1024 in a 1024 x 1 thread block
+    partialSum_smem[hipThreadIdx_x] = 0;                            // initialization of Shared to 0 using all 1024 x 1 threads
+
+    if (id_x >= xBufferLength)
+    {
+        return;
+    }
+
+    int xAlignedLength = xBufferLength & ~7;                        // alignedLength for vectorized global loads
+    int xDiff = xBufferLength - xAlignedLength;                     // difference between bufferLength and alignedLength
+    uint srcIdx = (id_z * xBufferLength) + id_x;
+
+    d_uint8 src_ui8;
+    *reinterpret_cast<d_uint8_s *>(&src_ui8) = *reinterpret_cast<d_uint8_s *>(srcPtr + srcIdx);
+
+    if (id_x + 8 > xBufferLength)
+        for(int i = xDiff; i < 8; i++)
+            src_ui8.ui1[i] = 0;                                     // local memory reset of invalid values (from the vectorized global load) to 0
+
+    src_ui8.ui4[0] += src_ui8.ui4[1];                               // perform small work of vectorized uint4 addition
+    partialSum_smem[hipThreadIdx_x] += (src_ui8.ui1[0] +
+                                        src_ui8.ui1[1] +
+                                        src_ui8.ui1[2] +
+                                        src_ui8.ui1[3]);            // perform small work of reducing uint32s to uint using 1024 x 1 threads and store in Shared
+    __syncthreads();                                                // syncthreads after Shared load
+
+    // Reduction of 1024 uints on 1024 threads per block in x dimension
+    for (int threadMax = 512; threadMax >= 1; threadMax /= 2)
+    {
+        if (hipThreadIdx_x < threadMax)
+            partialSum_smem[hipThreadIdx_x] += partialSum_smem[hipThreadIdx_x + threadMax];
+        __syncthreads();
+    }
+
+    // Final store to dst
+    if (hipThreadIdx_x == 0)
+    {
+        int totalElements = roiTensorPtrSrc[hipBlockIdx_z].xywhROI.roiHeight * roiTensorPtrSrc[hipBlockIdx_z].xywhROI.roiWidth;
+        dstPtr[hipBlockIdx_z] = (float)partialSum_smem[0] / totalElements;
+    }
+}
+
+__global__ void tensor_mean_grid_result_hip(Rpp32s *srcPtr,
```

Comment:
Both the functions for Rpp32s and Rpp32u are very much same
Is it possible to combine these into single function?

Reply:
Yes, spent some cycles checking to see if these can be combined but its not possible. We migth need to convert everything to float then use a templated kernel, For Rpp32s internally we use d_int8 and Rpp32u we use d_uint8 respectively.

Reply:
Only this portion of code is different as far as i understand
```
d_uint8 src_ui8;
    *reinterpret_cast<d_uint8_s *>(&src_ui8) = *reinterpret_cast<d_uint8_s *>(srcPtr + srcIdx);

    if (id_x + 8 > xBufferLength)
        for(int i = xDiff; i < 8; i++)
            src_ui8.ui1[i] = 0;                                     // local memory reset of invalid values (from the vectorized global load) to 0

    src_ui8.ui4[0] += src_ui8.ui4[1];                               // perform small work of vectorized uint4 addition
    partialSum_smem[hipThreadIdx_x] += (src_ui8.ui1[0] +
                                        src_ui8.ui1[1] +
                                        src_ui8.ui1[2] +
                                        src_ui8.ui1[3]);            // perform small work of reducing uint32s to uint using 1024 x 1 threads and store in Shared
    __syncthreads();              
```

Can you move this to a single helper function and have if else for Rpp32s and Rpp32u. Like how you did for one of the recent PR's?